iree/turbine/kernel/wave/schedule_reordering.py

```diff
@@ -0,0 +1,435 @@
+    SetWavePrio,
+    Write,
+)
+from ..lang.global_symbols import *
+
+from .scheduling.schedule_enums import SchedulingType
+from .utils.symbol_utils import subs_idxc
+
+
+##############################################################
+# General graph helper functions
+##############################################################
+
+"""
+Formatting for schedule reordering strategies:
+Values: 0xAB where:
+* A = Strategy types:
+  * 0 = No reordering
+  * 1 = Vanilla reordering
+  * 2 = Ping Pong reordering
+* B enumerates different strategy that share the same 0xA* bits.
+"""
+
+
+class SchedReorderStrategy(Enum):
+    NONE = 0x00
+    TWO_PP_CLUSTER = 0x20
+
+
+def get_graph_node(custom: CustomOp, graph: fx.Graph) -> fx.Node:
+    custom.add_to_graph(graph)
+    custom = custom.fx_node
+    return custom
+
+
+def has_single_root_mma(mma_nodes):
+    # Check only has single M, N, K dimensions.
+    indexing_dims = set(
+        [tuple(get_custom(mma_node).indexing_dims) for mma_node in mma_nodes]
+    )
+    if len(indexing_dims) != 1:
+        return False
+
+    # Check each MMA has unique expand dims => all come from a single non-expanded MMAOp.
+    hash_expand_dim = lambda x: tuple([(k, v) for k, v in x.expanded_dims.items()])
+    reduction_dims = set(
+        [hash_expand_dim(get_custom(mma_node)) for mma_node in mma_nodes]
+    )
+    if len(reduction_dims) != len(mma_nodes):
+        return False
+
+    # Check only a single reduction dim.
+    reduction_dims = set([get_custom(mma_node).reduction_dim for mma_node in mma_nodes])
+    if len(reduction_dims) != 1:
+        return False
+    return True
+
+
+def get_mma_tile_size(mma_nodes, constraints):
+    m_dims = set()
+    n_dims = set()
+    k_dims = set()
+    for mma_node in mma_nodes:
+        custom = get_custom(mma_node)
+        lhs_dim = set(get_custom(custom.lhs).indexing_dims)
+        rhs_dim = set(get_custom(custom.rhs).indexing_dims)
+        acc_dim = set(get_custom(custom.acc).indexing_dims)
+
+        m_dims.update(list(lhs_dim - rhs_dim))
+        n_dims.update(list(rhs_dim - lhs_dim))
+        # Subtract by acc dim to remove batch dims.
+        k_dims.update(list(rhs_dim.intersection(lhs_dim) - acc_dim))
+    if len(m_dims) != 1 or len(n_dims) != 1 or len(k_dims) != 1:
+        return None, None, None
+    mnk_dim = [list(dim)[0] for dim in (m_dims, n_dims, k_dims)]
+    mnk_tile = [
+        subs_idxc(tile_sym) for tile_sym in get_constrained_shape(mnk_dim, constraints)
+    ]
+    return mnk_tile
+
+
+def schedule_nodes_to_graph(graph, node_map, nodes):
+    for node in nodes:
+        custom = get_custom(node)
+        new_node = custom.copy(
+            new_graph=graph, arg_transform=lambda x: node_map[x] if x in node_map else x
+        )
+        node_map[node] = new_node.fx_node
+
+
+def reoreder_graph(graph, clusters):
+    node_list = list(graph.nodes)
+    prune_duplicates = lambda x: list(dict.fromkeys(x))
+    prune_reordering_nodes = lambda x: [
```

Comment:
What is this lambda doing? Could you add some documentation either here or on the rst page?